typings/index.d.ts

```diff
@@ -779,33 +781,148 @@ export class GuildAuditLogs {
+    ? 'EMOJI'
+    : TAction extends 'MESSAGE_DELETE' | 'MESSAGE_BULK_DELETE' | 'MESSAGE_PIN' | 'MESSAGE_UNPIN'
+    ? 'MESSAGE'
+    : TAction extends 'INTEGRATION_CREATE' | 'INTEGRATION_UPDATE' | 'INTEGRATION_DELETE'
+    ? 'INTEGRATION'
+    : TAction extends 'STAGE_INSTANCE_CREATE' | 'STAGE_INSTANCE_UPDATE' | 'STAGE_INSTANCE_DELETE'
+    ? 'STAGE_INSTANCE'
+    : TAction extends 'STICKER_CREATE' | 'STICKER_UPDATE' | 'STICKER_DELETE'
+    ? 'STICKER'
+    : TAction extends 'THREAD_CREATE' | 'THREAD_UPDATE' | 'THREAD_DELETE'
+    ? 'THREAD'
+    : 'UNKNOWN',
+> {
   public constructor(logs: GuildAuditLogs, guild: Guild, data: RawGuildAuditLogEntryData);
-  public action: GuildAuditLogsAction;
-  public actionType: GuildAuditLogsActionType;
+  public action: TAction;
+  public actionType: TActionType;
   public changes: AuditLogChange[] | null;
   public readonly createdAt: Date;
   public readonly createdTimestamp: number;
   public executor: User | null;
-  public extra: unknown | Role | GuildMember | null;
+  public extra: TAction extends 'MEMBER_PRUNE'
+    ? { removed: number; days: number }
+    : TAction extends 'MEMBER_MOVE' | 'MESSAGE_DELETE' | 'MESSAGE_BULK_DELETE'
+    ? { channel: GuildChannel | { id: Snowflake }; count: number }
+    : TAction extends 'MESSAGE_PIN' | 'MESSAGE_UNPIN'
+    ? { channel: GuildChannel | { id: Snowflake }; messageId: Snowflake }
+    : TAction extends 'MEMBER_DISCONNECT'
+    ? { count: number }
+    : TAction extends 'CHANNEL_OVERWRITE_CREATE' | 'CHANNEL_OVERWRITE_UPDATE' | 'CHANNEL_OVERWRITE_DELETE'
+    ?
+        | Role
+        | GuildMember
+        | { id: Snowflake; name: string; type: OverwriteTypes.role }
+        | { id: Snowflake; type: OverwriteTypes.member }
+    : TAction extends 'STAGE_INSTANCE_CREATE' | 'STAGE_INSTANCE_DELETE' | 'STAGE_INSTANCE_UPDATE'
+    ? GuildChannel | { id: Snowflake }
+    : null;
   public id: Snowflake;
   public reason: string | null;
-  public target:
-    | Guild
-    | GuildChannel
-    | User
-    | Role
-    | GuildEmoji
-    | Invite
-    | Webhook
-    | Message
-    | Integration
-    | StageInstance
-    | Sticker
-    | ThreadChannel
-    | { id: Snowflake }
-    | null;
-  public targetType: GuildAuditLogsTarget;
+  public target: TTargetType extends 'USER'
+    ? User | null
+    : TTargetType extends 'GUILD'
+    ? Guild
+    : TTargetType extends 'WEBHOOK'
+    ? Webhook
+    : TTargetType extends 'INVITE'
+    ? Promise<Invite> | { [x: string]: unknown }
+    : TTargetType extends 'MESSAGE'
+    ? TActionType extends 'MESSAGE_BULK_DELETE'
+      ? Guild | { id: Snowflake }
+      : User
+    : TTargetType extends 'INTEGRATION'
+    ? Integration
+    : TTargetType extends 'CHANNEL' | 'THREAD'
+    ? GuildChannel | ThreadChannel | { id: Snowflake; [x: string]: unknown }
+    : TTargetType extends 'STAGE_INSTANCE'
+    ? StageInstance
+    : TTargetType extends 'STICKER'
+    ? Sticker
+    : Role | GuildEmoji | { id: Snowflake } | null;
```

Comment:
Instead of using long conditional type chains, you should consider using a mapped type instead where `TYPE` is mapped to `ObjectType` https://www.typescriptlang.org/docs/handbook/2/mapped-types.html

Then easily reference the varient type later on:
`MyMappedType['USER'] // Returns User Type`

This reduces the work that the typechecker needs to do, and makes it less verbose.

Reply:
An example can be found in this diff: https://github.com/discordjs/discord.js/pull/6476

Reply:
Done in https://github.com/noftaly/discord.js/commit/70a4b72b2f30a4eb037639b21c0f8e958d69ba2c